### PR TITLE
feat(fields): Definitions and refactoring for aggregation fields

### PR DESCRIPTION
**Pull request by jennmueng**
Define definitions for aggregation functions + refactor uses of them to use new definitions from the consolidated field definitions.

Old string literal union type `AggregationKey` that contains aliases is renamed to `AggregationKeyWithAlias` while the new `AggregationKey` is an enum.

### CODE REVIEW

File: static/app/utils/discover/fields.tsx [hunk 1]
```diff
@@ -483,35 +465,31 @@ export const AGGREGATIONS = {
...
-  tps: 'eps',
+  tpm: AggregationKey.Epm,
+  tps: AggregationKey.Eps,
 };
 
-assert(AGGREGATIONS as Readonly<{[key in keyof typeof AGGREGATIONS]: Aggregation}>);
+assert(AGGREGATIONS as Readonly<{[key in AggregationKey]: Aggregation}>);
 
-export type AggregationKey = keyof typeof AGGREGATIONS | keyof typeof ALIASES | '';
+export type ValidAggregationKey = `${AggregationKey}` | keyof typeof ALIASES | '';
```
Comment (jennmueng) [inline static/app/utils/discover/fields.tsx:492]: Did not want to completely refactor the whole codebase with who knows how many string literals are used for the aggregation names, so `ValidAggregationKey` is a stopgap so existing aggregation definitions will still work without linking or ts errors.

File: static/app/components/smartSearchBar/utils.tsx [hunk 2]
```diff
@@ -351,6 +351,15 @@ export const shortcuts: Shortcut[] = [
   },
 ];
 
+const getItemTitle = (key: string, kind: FieldKind) => {
+  if (kind === FieldKind.FUNCTION) {
+    // Replace the function innards with ... for cleanliness
+    return key.replace(/(?<=\()[^\)]*/g, '...');
```
Comment (evanpurkhiser) [inline static/app/components/smartSearchBar/utils.tsx:357]: [Is that a negative lookbehind?](https://caniuse.com/js-regexp-lookbehind)

I don't think you can use the lookbehind in safari 😢

File: static/app/components/smartSearchBar/utils.tsx [hunk 3]
```diff
@@ -367,12 +376,16 @@ export const getTagItemsFromKeys = (
     .reduce((groups, key) => {
       const keyWithColon = `${key}:`;
       const sections = key.split('.');
-      const definition = getFieldDefinition(key);
+
+      const definition =
+        supportedTags[key]?.kind === FieldKind.FUNCTION
+          ? getFieldDefinition(key.split('(')[0])
```
Comment (evanpurkhiser) [inline static/app/components/smartSearchBar/utils.tsx:382]: kinda wish this was a bit more elegant but 🤷

File: static/app/utils/fields/index.ts [hunk 4]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: null,
+  },
+  [AggregationKey.UserMisery]: {
+    desc: 'Returns transactions of miserable users between 0 to 1',
+    kind: FieldKind.FUNCTION,
+    valueType: null,
+  },
+  [AggregationKey.LastSeen]: {
+    desc: 'Issues last seen at a date and time',
```
Comment (evanpurkhiser) [inline static/app/utils/fields/index.ts:281]: oh were you gonna wrap these with `t()`

**Review by evanpurkhiser — APPROVED** [review]

File: static/app/components/smartSearchBar/utils.tsx (hunk 2 above)
Comment (jennmueng) [reply static/app/components/smartSearchBar/utils.tsx]: you're right

Comment (evanpurkhiser) [reply static/app/components/smartSearchBar/utils.tsx]: I only know because I brought all of sentry down once with this

Comment (jennmueng) [reply static/app/components/smartSearchBar/utils.tsx]: oof

**Comment by jennmueng** [comment]
ok tried on safari works now lol

File: static/app/utils/fields/index.ts [hunk 5]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.NUMBER,
+  },
+  [AggregationKey.CountUnique]: {
+    desc: t('Returns the unique count for the unique field'),
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.INTEGER,
+  },
+  [AggregationKey.CountMiserable]: {
+    desc: t('Returns a score of miserable users'),
```
Comment (wmak) [inline static/app/utils/fields/index.ts:181]:
```suggestion
    desc: t('Count of unique miserable users'),
```

File: static/app/utils/fields/index.ts [hunk 6]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.NUMBER,
+  },
+  [AggregationKey.CountIf]: {
+    desc: t('Returns results based on conditional case'),
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.NUMBER,
+  },
+  [AggregationKey.CountWebVitals]: {
+    desc: t('Returns count of web vitals with a specific status'),
```
Comment (wmak) [inline static/app/utils/fields/index.ts:191]:
```suggestion
    desc: t('Count of web vitals with a specific status'),
```

File: static/app/utils/fields/index.ts [hunk 7]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.NUMBER,
+  },
+  [AggregationKey.Eps]: {
+    desc: t('Returns events per second'),
```
Comment (wmak) [inline static/app/utils/fields/index.ts:196]:
```suggestion
    desc: t('Events per second'),
```

File: static/app/utils/fields/index.ts [hunk 8]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.NUMBER,
+  },
+  [AggregationKey.Epm]: {
+    desc: t('Returns events per minute'),
```
Comment (wmak) [inline static/app/utils/fields/index.ts:201]:
```suggestion
    desc: t('Events per minute'),
```

File: static/app/utils/fields/index.ts [hunk 9]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.NUMBER,
+  },
+  [AggregationKey.FailureRate]: {
+    desc: t('Returns percentage of failed transactions'),
+    kind: FieldKind.FUNCTION,
+    valueType: FieldValueType.PERCENTAGE,
+  },
+  [AggregationKey.FailureCount]: {
+    desc: t('Returns failed event count based on transaction.status'),
```
Comment (wmak) [inline static/app/utils/fields/index.ts:211]:
```suggestion
    desc: t('Failed event count based on transaction.status'),
```

File: static/app/utils/fields/index.ts [hunk 10]
```diff
@@ -133,13 +133,157 @@ export enum MobileVital {
...
+    kind: FieldKind.FUNCTION,
+    valueType: null,
+  },
+  [AggregationKey.Sum]: {
+    desc: t('Returns the total value for the selected field'),
+    kind: FieldKind.FUNCTION,
+    valueType: null,
+  },
+  [AggregationKey.Any]: {
+    desc: t('Transactions with a specific duration'),
```
Comment (wmak) [inline static/app/utils/fields/index.ts:231]:
```suggestion
    desc: t('Not Recommended, a random field value'),
```

**Pull request merged**